src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,123 @@
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "random/generators.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/defer.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+#include <string>
+
+namespace cloud_storage {
+
+cache::cache(std::filesystem::path cache_dir) noexcept
+  : _cache_dir(std::move(cache_dir))
+  , _cnt(0) {}
+
+ss::future<> cache::start() {
+    vlog(cst_log.debug, "Starting archival cache service");
+    // todo: start maintenance cycle here
+    return ss::make_ready_future<>();
+}
+
+ss::future<> cache::stop() {
+    vlog(cst_log.debug, "Stopping archival cache service");
+    return _gate.close();
+}
+
+ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    gate_guard guard{_gate};
+    vlog(cst_log.debug, "Trying to get {} from archival cache.", key.native());
+    ss::file cache_file;
+    try {
+        cache_file = co_await ss::open_file_dma(
+          (_cache_dir / key).native(), ss::open_flags::ro);
+    } catch (...) {
+        vlog(
+          cst_log.debug,
+          "Couldn't open file {} to read: {}",
+          (_cache_dir / key).native(),
+          std::current_exception());
+        co_return std::nullopt;
+    }
+
+    auto data_size = co_await cache_file.size();
+    auto data_stream = ss::make_file_input_stream(cache_file);
+    co_return std::optional(cache_item{std::move(data_stream), data_size});
+}
+
+ss::future<>
+cache::put(std::filesystem::path key, ss::input_stream<char> data) {
+    gate_guard guard{_gate};
+    vlog(cst_log.debug, "Trying to put {} to archival cache.", key.native());
+
+    auto filename = (_cache_dir / key).filename();
+    auto dir_path = (_cache_dir / key).remove_filename();
+    co_await ss::recursive_touch_directory(dir_path.string());
+
+    // tmp file is used to protect against concurrent writes to the same file.
+    // One tmp file is written only once by one thread. tmp file should not be
+    // read directly. _cnt is an atomic counter that ensures the uniqueness of
+    // names for tmp files.
+    auto tmp_filename = std::filesystem::path(
+      filename.string() + std::to_string(++_cnt) + ".part");
+    auto flags = ss::open_flags::wo | ss::open_flags::create
+                 | ss::open_flags::exclusive;
+    auto tmp_cache_file = co_await ss::open_file_dma(
+      (dir_path / tmp_filename).native(), flags);
+    auto out = co_await ss::make_file_output_stream(tmp_cache_file);
+
+    co_await ss::copy(data, out)
+      .then([&out]() { return out.flush(); })
+      .finally([&out, &data]() {
+          return ss::when_all_succeed(data.close(), out.close());
```

Comment:
`data` is closed a bit inconsistently - e.g. if there is an exception when opening the file it is left open, but if there is an exception when writing to `out`, it is closed. Anyway, it necessary to close `data` here? I think that should be the responsibility of the caller - who created it must close it.